### PR TITLE
Fix incompatibilities with slam_toolbox

### PR DESCRIPTION
**Pull request by justinIRBT**
- Fix angle compensate mode to publish angle compensated values
- Fix angle_increment calculation
- Add optional flip_x_axis option to deal with issue discussed here: https://github.com/SteveMacenski/slam_toolbox/issues/198.  Flip x-axis can be used when laser is mounted with motor behind it as rotated TF laser frame doesn't seem to work with slam_toolbox.

### CODE REVIEW

**Comment by chrisl8** [comment]
@allenh1 Are you planning to merge this? It is kind of now required if anybody wants to use slam_toolbox, since the matching PR was merged over there.

This also brings up the issue of the ROS1 Rplidar driver. This is also required to make that work, but I don't think Slamtec is going to merge anything anytime soon.
Do you have any plans to host a patched ROS1 Rplidar driver too?

**Comment by justinIRBT** [comment]
I suspect the Slamtech ROS1 driver will work with the slam toolbox change if the RPLIDAR is mounted X axis forward (motor in front).  In https://github.com/Slamtec/rplidar_ros/blob/master/src/node.cpp, the angle_increment is correctly calculated with:
```
    scan_msg.angle_increment =
        (scan_msg.angle_max - scan_msg.angle_min) / (double)(node_count-1);
```
and the angle_compensated_nodes are correctly published:
```
publish_scan(&scan_pub, angle_compensate_nodes, angle_compensate_nodes_count,
                             start_scan_time, scan_duration, inverted,
                             angle_min, angle_max, max_distance,
                             frame_id);
```

**Comment by allenh1** [comment]
> @allenh1 Are you planning to merge this?

Yes, after I gave time to test things

> Do you have any plans to host a patched ROS1 Rplidar driver too?

I feel like I'm stepping on toes already, to be honest. If I did that I feel like I'd be stomping on them, haha.

That said, if they are ignoring good PR's, I'll be open to making a noetic branch as well.

File: src/rplidar_node.cpp
Comment (allenh1) [inline src/rplidar_node.cpp:161]: @chrisl8 wasn't this changed to be correct the other way? I'm confused now

Comment (justinIRBT) [reply src/rplidar_node.cpp:161]: I was working with angle_compensated mode, in slam_toolbox, karto needs a consistent number of scan readings to be happy, the non-angle_compensated mode was giving variable readings each iteration.  Angle_compensated is giving 1 degree readings from [0,359] (turned into [-179,180]), so 360 readings.  Without the (node_count-1), it is reporting the angle increment (using degrees for readability) as 359/360 degrees, as opposed to the 1 degree expected.  I am not sure if that was a work-around for the threshold issue I fixed in slam_toolbox, as Karto is doing:
```
m_NumberOfRangeReadings = static_cast<kt_int32u>(math::Round((GetMaximumAngle() -
      GetMinimumAngle()) /
      GetAngularResolution()) + residual)
```
Where residual defaults to 1, unless a laser is reported to Karto as a 360 laser, but this is generally intended when a laser goes from [-PI,PI] and you don't want to count the beginning and end reading twice.  Because angle_compensated mode is reporting as [-179,180], it is not wrapping around, but the previous slam_toolbox check for 360 laser was doing ((max_angle - min_angle) - 2PI) < 0.1, which was incorrectly reporting [-179,180] as a 360 Lidar, so that led residual to be 0 in the equation above and would have led Karto to expect 359 readings.  Therefore, I can see making angle_increment report as 359/360 would have led the Karto math to report the expected 360 readings.  Now that slam_toolbox is fixed, this laser driver can use the correct angle_increment reporting 359/359 and the residual on Karto is 1 and it expects 360 readings.

Comment (chrisl8) [reply src/rplidar_node.cpp:161]: I have not tested all of this yet, but my understanding was that @justinIRBT created a superior fix in place of my hack.

He was also able to convince the slam_toolbox maintainer to alter their code, which I was unable to do. Probably because my explanation was poor and I didn't submit a PR.

I will have to back port this to the ROS1 driver before I can test it.

**Comment by chrisl8** [comment]
I wouldn't personally worry too much about "stepping on toes." I could host a ROS1 driver too, but I like having it in one place, and while I can patch together code and test things, I'm over my head in the math and theory about this stuff.
In the open source world, there are often multiple sources of "truth", and at the end of the day, there is not actually any reason Slamtec has to maintain a ROS driver for their hardware. They are perfectly within their rights to sell hardware and let others build the software. So it is a two way street.
Besides, if we keep something maintained here that is well reviewed and works, we can ask them to merge it into theirs if they ever start updating their repo again.

One thing that bothers me is that the file names are now different between ROS1 and ROS2 drivers, making it hard to use the same commits for both.
It would be nice to update the ROS1 driver to look like the ROS2 driver, which I'm sure Slamtec isn't going to do.

When I get time, I'll try creating a noetic targeted driver build and make a PR and then you can respond as you see fit.

**Comment by allenh1** [comment]
> One thing that bothers me is that the file names are now different between ROS1 and ROS2 drivers, making it hard to use the same commits for both.
It would be nice to update the ROS1 driver to look like the ROS2 driver, which I'm sure Slamtec isn't going to do.

Yeah, that is true. I'm likely going to remove the old driver soon in favor of the component implementation so that it doesn't confuse anyone.

File: src/rplidar_node.cpp
Comment (allenh1) [inline src/rplidar_node.cpp]:
```suggestion
  for (size_t i = 0; i < node_count; ++i) {
```

Nit: I prefer pre-increment

**Review by allenh1 — CHANGES_REQUESTED** [review]
Please fix the white space, but, other than that, this changeset looks fine to me.

I've been using `ament_uncrustify` for the linter.

File: src/rplidar_node.cpp
Comment (justinIRBT) [reply src/rplidar_node.cpp]: The diff is a little messy, but its just converting the 4 spaces to 2 and switching the increment you mentioned

File: src/rplidar_node.cpp
Comment (chrisl8) [reply src/rplidar_node.cpp:161]: FYI: I confirmed just now in testing that this update to slam_toolbox works with the ROS1 RPlidar driver provided by Slamtec.
That doesn't really have any affect on your merging of this PR for the ROS2 driver, but it does mean that updating the ROS1 driver isn't highly important.

**Review by allenh1 — APPROVED** [review]
lgtm -- thanks for the patch!

**Pull request merged**
